smithy-model/src/main/java/software/amazon/smithy/model/shapes/Shape.java

```diff
@@ -88,10 +89,43 @@
         }
     }
 
-    protected void validateMixins(Collection<ShapeId> mixins) {
-        // This method is overridden in subclasses that allow mixins (structure, union, member).
-        // This situation can only happen by manipulating the semantic model in code.
-        throw new SourceException("Mixins are not allowed on " + getId() + ", a " + getType(), this);
+    protected void validateMixins(Map<ShapeId, Shape> mixins, Map<ShapeId, Trait> introducedTraits) {
+        Set<String> invalid = new TreeSet<>();
+        for (Shape mixin : mixins.values()) {
+            if (mixin.getType() != getType()) {
+                invalid.add(mixin.getId().toString());
+            }
+        }
+        if (!invalid.isEmpty()) {
+            String invalidList = String.join("`, `", invalid);
+            throw new IllegalStateException(String.format(
+                    "Mixins may only be mixed into shapes of the same type. The following mixins were applied to the "
+                            + "%s shape `%s` which are not %1$s shapes: [`%s`]", getType(), getId(), invalidList));
+        }
+    }
+
+    protected MemberShape getRequiredMixinMember(
+            AbstractShapeBuilder<?, ?> builder,
+            String name
+    ) {
+        MemberShape mixedMember = null;
```

Comment:
Is this method called in a way that it should return any members set locally on the builder vs members of mixins?

Reply:
`AbstractShapeBuilder` doesn't have members, so the callers have to supply the local version if present.